framework_laptop.c

```diff
@@ -257,6 +258,357 @@ static int framework_laptop_battery_remove(struct power_supply *battery)
+	if (sen_attr->index != 0) {
+		return -EINVAL;
+	}
+
+	u32 val;
+	if (ec_get_target_rpm(sen_attr->index, &val) < 0) {
+		return -EIO;
+	}
+
+	// Format as string for sysfs
+	return sysfs_emit(buf, "%u\n", val);
+}
+
+// --- fanN_fault ---
+static ssize_t fw_fan_fault_show(struct device *dev,
+				 struct device_attribute *attr, char *buf)
+{
+	struct sensor_device_attribute *sen_attr = to_sensor_dev_attr(attr);
+
+	u16 val;
+	if (ec_get_fan_speed(sen_attr->index, &val) < 0) {
+		return -EIO;
+	}
+
+	// Format as string for sysfs
+	return sysfs_emit(buf, "%u\n", val == EC_FAN_SPEED_NOT_PRESENT);
+}
+
+// --- fanN_alarm ---
+static ssize_t fw_fan_alarm_show(struct device *dev,
+				 struct device_attribute *attr, char *buf)
+{
+	struct sensor_device_attribute *sen_attr = to_sensor_dev_attr(attr);
+
+	u16 val;
+	if (ec_get_fan_speed(sen_attr->index, &val) < 0) {
+		return -EIO;
+	}
+
+	// Format as string for sysfs
+	return sysfs_emit(buf, "%u\n", val == EC_FAN_SPEED_STALLED);
+}
+
+// --- pwmN_enable ---
+static ssize_t ec_set_auto_fan_ctrl(u8 idx)
+{
+	int ret;
+	if (!ec_device)
+		return -ENODEV;
+
+	struct cros_ec_device *ec = dev_get_drvdata(ec_device);
+
+	struct ec_params_auto_fan_ctrl_v1 params = {
+		.fan_idx = idx,
+	};
+
+	ret = cros_ec_cmd(ec, 1, EC_CMD_THERMAL_AUTO_FAN_CTRL, &params,
+			  sizeof(params), NULL, 0);
+	if (ret < 0)
+		return -EIO;
+
+	return 0;
+}
+
+static ssize_t fw_pwm_enable_store(struct device *dev,
+				   struct device_attribute *attr,
+				   const char *buf, size_t count)
+{
+	struct sensor_device_attribute *sen_attr = to_sensor_dev_attr(attr);
+
+	// The EC doesn't take any arguments for this command,
+	// so we don't need to parse the buffer
+	// u32 val;
+
+	// int err;
+	// err = kstrtou32(buf, 10, &val);
+	// if (err < 0)
+	// 	return err;
+
+	if (ec_set_auto_fan_ctrl(sen_attr->index) < 0) {
+		return -EIO;
+	}
+
+	return count;
+}
+
+// --- pwmN ---
+static ssize_t ec_set_fan_duty(u8 idx, u32 *val)
+{
+	int ret;
+	if (!ec_device)
+		return -ENODEV;
+
+	struct cros_ec_device *ec = dev_get_drvdata(ec_device);
+
+	struct ec_params_pwm_set_fan_duty_v1 params = {
+		.percent = *val,
+		.fan_idx = idx,
+	};
+
+	ret = cros_ec_cmd(ec, 1, EC_CMD_PWM_SET_FAN_DUTY, &params,
+			  sizeof(params), NULL, 0);
+	if (ret < 0)
+		return -EIO;
+
+	return 0;
+}
+
+static ssize_t fw_pwm_store(struct device *dev, struct device_attribute *attr,
+			    const char *buf, size_t count)
+{
+	struct sensor_device_attribute *sen_attr = to_sensor_dev_attr(attr);
+	u32 val;
+
+	int err;
+	err = kstrtou32(buf, 10, &val);
+	if (err < 0)
+		return err;
+
+	if (ec_set_fan_duty(sen_attr->index, &val) < 0) {
+		return -EIO;
+	}
+
+	return count;
+}
+
+static ssize_t fw_pwm_min_show(struct device *dev,
+			       struct device_attribute *attr, char *buf)
+{
+	return sysfs_emit(buf, "%i\n", 0);
+}
+
+static ssize_t fw_pwm_max_show(struct device *dev,
+			       struct device_attribute *attr, char *buf)
+{
+	return sysfs_emit(buf, "%i\n", 100);
```

Comment:
Just for my understanding - all the PWM values are duty cycle percentages, right? Is that by convention, or specific to this driver? I don't understand hwmon all that well 😅

Reply:
No worries idk much either, but a lot of other stuff seems to be 0-255, so I sorta just wanted to highlight that this only takes 0-100.